extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs

```diff
@@ -57,136 +59,166 @@ private async ValueTask<ConversionResult> ConvertFromBindingDataAsync(ConverterC
+                Constants.JsonContentType => bindingData?.Content?.ToObjectFromJson<BlobBindingData>(),
+                _ => throw new InvalidContentTypeException(Constants.JsonContentType)
+            };
+        }
 
-                    if (modelBindingData.Source is not Constants.BlobExtensionName)
-                    {
-                        throw new InvalidBindingSourceException(Constants.BlobExtensionName);
-                    }
+        private async Task<object?> ConvertModelBindingDataAsync(Type targetType, BlobBindingData blobData)
+        {
+            if (string.IsNullOrEmpty(blobData.Connection))
+            {
+                throw new ArgumentNullException(nameof(blobData.Connection));
+            }
 
-                    Dictionary<string, string> content = GetBindingDataContent(modelBindingData);
-                    var element = await ConvertModelBindingDataAsync(content, elementType, modelBindingData);
+            if (string.IsNullOrEmpty(blobData.ContainerName))
+            {
+                throw new ArgumentNullException(nameof(blobData.ContainerName));
+            }
 
-                    if (element is not null)
-                    {
-                        result[i] = element;
-                    }
-                }
+            BlobContainerClient container = CreateBlobContainerClient(blobData.Connection!, blobData.ContainerName!);
 
-                if (!context.TargetType.IsArray)
+            if (IsCollectionBinding(targetType, blobData.BlobName!))
+            {
+                return await BindToCollectionAsync(targetType, container, blobData.BlobName!);
+            }
+            else
+            {
+                if (targetType == typeof(BlobContainerClient) && !string.IsNullOrEmpty(blobData.BlobName))
                 {
-                    var resultType = typeof(List<>).MakeGenericType(elementType);
-                    result = (IList)Activator.CreateInstance(resultType, result);
+                    throw new InvalidOperationException("Binding to BlobContainerClient with a blob path is not supported. "
+                                                        + "Either bind to the container path, or use BlobClient instead.");
                 }
 
-                return ConversionResult.Success(result);
+                return await ToTargetTypeAsync(targetType, container, blobData.BlobName!);
             }
-            catch (JsonException ex)
+        }
+
+        /// <summary>
+        /// Determines if the binding is a collection binding.
+        /// A collection binding is when the target type is an array or IEnumerable and
+        /// the blob name is either null or empty (meaning a container path is provided).
+        /// If a blob name is provided, it must be a directory path (no file extension provided).
+        /// </summary>
+        private bool IsCollectionBinding(Type type, string blobName)
+        {
+            if (type == typeof(string) || type == typeof(byte[]))
```

Comment:
Trying to remember, didn't we support binding a blob to `string[]` to support the deserialization case? Would be good to have a test validating that, as I don't think this will currently support that flow.

Reply:
Using `string[]` as the example, but the concern here applies to any JSON array deserialization target.